Time conversion histogram

liyiy opened the pull request:
## Changes

*Please describe.*  
*If this affects the frontend, include screenshots.*  

Functionality for displaying time to convert histogram and changing steps

<img width="551" alt="Screen Shot 2021-07-09 at 8 06 24 PM" src="https://user-images.githubusercontent.com/25164963/125146294-16da1d80-e0f3-11eb-9173-d5d1080c2c51.png">

https://user-images.githubusercontent.com/25164963/125139654-ed16fb80-e0de-11eb-9296-734ab351ab9b.mov


## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

liyiy committed:
set up funnel histogram option

liyiy committed:
connect time conversion bins backend with frontend logic

liyiy committed:
typing

liyiy committed:
add histogram step change functionality

liyiy committed:
clean up a little

liyiy commented:
1. The histogram doesn't work for `all steps` because it hasn't been implemented in the backend yet. 
2. I have to fix the step dropdown to include more information
3. I have to clean up the header bar 
4. This will probably be in another PR -> do we want to continue using our existing chartjs bar component implementation and just make style changes to it to look more like the mocks, or create something new for it?

liyiy committed:
header style updates

liyiy committed:
add new color

liyiy committed:
update step dropdown select

liyiy committed:
move component variables to kea and funnel utility methods into separate file

paolodamico committed:
Merge branch 'master' into time-conversion-histogram

paolodamico commented on frontend/src/lib/constants.tsx:
```diff
@@ -11,6 +11,7 @@ export const ACTIONS_BAR_CHART = 'ActionsBar'
 export const ACTIONS_BAR_CHART_VALUE = 'ActionsBarValue'
 export const PATHS_VIZ = 'PathsViz'
 export const FUNNEL_VIZ = 'FunnelViz'
+export const FUNNELS_HISTOGRAM = 'FunnelsHistogram'
```
Nit: In case we later change the type of visualisation, something like `FunnelTimeViz` might work better?

paolodamico committed:
adjust graph type selector

paolodamico commented on frontend/src/types.ts:
```diff
@@ -605,6 +606,8 @@ export interface FilterType {
     filter_test_accounts?: boolean
     from_dashboard?: boolean
     funnel_step?: number
+    funnel_viz_type?: string
```
I think this can be typed in more detail, e.g. `time_to_convert`

paolodamico commented on frontend/src/types.ts:
```diff
@@ -605,6 +606,8 @@ export interface FilterType {
     from_dashboard?: boolean
     funnel_step?: number
+    funnel_viz_type?: string
+    funnel_to_step?: number
```
Worth adding a note clarifying what this is used for

paolodamico committed:
fix colors

paolodamico commented on frontend/src/scenes/funnels/funnelLogic.ts:
```diff
@@ -57,7 +75,7 @@ export const cleanFunnelParams = (filters: FilterType): FilterType => {
     [...(filters.actions || []), ...(filters.events || [])].length === 0
 
-export const funnelLogic = kea<funnelLogicType>({
+export const funnelLogic = kea<funnelLogicType<TimeStepOption>>({
```
I think the new kea typegen no longer requires injecting deps?

liyiy replied:
hmm I think my file just added this by itself, I didn't manually put this in, should I still take it out or is this part of kea magic? :o

paolodamico reviewed (approved):
Code lgtm! Some minor details but they can be fixed in a follow up PR. Couldn't fully test functionality as I'm having problems with local CH, but we can test and iterate online as it's behind a FF.

paolodamico committed:
missing commit

mariusandra reviewed (commented):
Gave a quick look. I'd improve naming of a few things, and left some other inline comments.

liyiy commented on frontend/src/scenes/funnels/funnelLogic.ts:
```diff
@@ -49,6 +65,8 @@ export const cleanFunnelParams = (filters: FilterType): FilterType => {
         ...(filters.filter_test_accounts ? { filter_test_accounts: filters.filter_test_accounts } : {}),
         ...(filters.funnel_step ? { funnel_step: filters.funnel_step } : {}),
+        ...(filters.funnel_viz_type ? { funnel_viz_type: filters.funnel_viz_type } : {}),
+        ...(filters.funnel_step ? { funnel_to_step: filters.funnel_step } : {}),
```
yeah, i'll have to update this from the backend, these are just the params i was given 😄 but definitely.. `funnel_step` and `funnel_to_step`

mariusandra replied:
Well, sometimes you need to work with what you got. If that's too much to change, that's also fine, but then I'd add some `// NB! funky names` comment to warn fellow adventurers of dragons ahead.

neilkakkar replied:
If it helps, on the backend the mixin is actually called `FunnelPersonsStepMixIn`, since the `funnel_step` parameter is only used for querying parameters. But good call, think we should rename the key value itself to reflect that.

liyiy commented on frontend/src/types.ts:
```diff
@@ -682,7 +685,7 @@ export interface TrendResultWithAggregate extends TrendResult {
 export interface FunnelStep {
     action_id: string
-    average_time: number
+    average_conversion_time: number
```
yep, these params all come from the backend changes

liyiy commented on frontend/src/scenes/funnels/funnelLogic.ts:
```diff
@@ -128,7 +145,27 @@ export const funnelLogic = kea<funnelLogicType>({
                     breakpoint()
                     insightLogic.actions.endQuery(queryId, ViewType.FUNNELS, result.last_refresh)
-                    actions.setSteps(result.result)
+                    actions.setSteps(result.result as FunnelStep[])
+                    let binsResult: FunnelsTimeConversionResult
+                    if (params.display === ChartDisplayType.FunnelsHistogram) {
+                        try {
+                            params.funnel_viz_type = 'time_to_convert'
+                            params.funnel_to_step = values.histogramStep
+                            binsResult = await pollFunnel(params)
+                        } catch (e) {
+                            insightLogic.actions.endQuery(queryId, ViewType.FUNNELS, null, e)
+                            eventUsageLogic.actions.reportFunnelCalculated(
+                                eventCount,
+                                actionCount,
+                                interval,
+                                false,
+                                e.message
+                            )
+                            return []
+                        }
+                        insightLogic.actions.endQuery(queryId, ViewType.FUNNELS, binsResult.last_refresh)
+                        actions.setTimeConversionBins(binsResult.result as number[])
+                    }
```
I ended up having to make both API calls because for example the steps dropdown selection contains data that isn't found in the time conversion funnel call (average conversion time and conversion rate for the other steps). 

<img width="551" alt="Screen Shot 2021-07-09 at 8 06 24 PM" src="https://user-images.githubusercontent.com/25164963/125301277-d4822d80-e2f8-11eb-96c4-e24352ffe2b5.png">

mariusandra replied:
Ok, makes sense :).

liyiy replied:
that method can definitely be cleaned up though 😭 it's ginormous, shall be on the TODO

liyiy committed:
update naming

liyiy committed:
Merge branch 'time-conversion-histogram' of https://github.com/PostHog/posthog into time-conversion-histogram

liyiy committed:
remove dupe option

liyiy commented:
I'll merge this so @paolodamico can have some time to click around this before the user interview, I think I've addressed all the comments except for the confusing `funnel_viz_type` stuff, for which I've left a comment for and will update in a next PR since it touches the backend and I'd prefer to break things after this can be used for the user interview
